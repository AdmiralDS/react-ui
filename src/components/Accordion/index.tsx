import * as React from 'react';
import styled from 'styled-components';
import { typography } from '#src/components/Typography';
import { ReactComponent as ChevronDownOutline } from '@admiral-ds/icons/build/system/ChevronDownOutline.svg';
import { uid } from '#src/components/common/uid';
import { keyboardKey } from '#src/components/common/keyboardKey';

import { Collapse } from './Collapse';
import { moveFocus, nextItem, previousItem } from './utils';

type Dimension = 'l' | 'm';

const Chevron = styled(ChevronDownOutline)`
  transition: all 0.3s;
  flex-shrink: 0;
  width: 24px;
  height: 24px;
  margin-left: 8px;
  & *[fill^='#'] {
    fill: ${({ theme }) => theme.color.text.secondary};
  }
  [data-disabled='true'] & {
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.text.tertiary};
    }
  }
  [data-dimension='m'] & {
    width: 20px;
    height: 20px;
  }
`;

const ItemTitleContent = styled.span`
  position: relative;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: space-between;
  align-items: flex-start;

  padding: 16px 16px 15px 16px;
  [data-dimension='m'] && {
    padding: 10px 16px 9px 16px;
  }
`;

const ItemTitle = styled.button`
  position: relative;
  display: flex;
  justify-content: space-between;
  align-items: center;
  min-height: 54px;
  width: 100%;
  background: transparent;
  -webkit-tap-highlight-color: transparent;
  text-align: left;
  border: none;
  background: none;
  margin: 0;
  padding: 0;
  cursor: pointer;
  overflow: visible;
  ${typography['Subtitle/Subtitle 2']}
  &:hover {
    background: ${(p) => p.theme.color.background.secondary};
    & *[fill^='#'] {
      fill: ${(p) => p.theme.color.basic.hover};
    }
  }
  &:focus {
    &:before {
      position: absolute;
      content: '';
      border: 2px solid ${({ theme }) => theme.color.basic.hover};
      top: -1px;
      left: 0;
      bottom: -1px;
      right: 0;
    }
  }
  &:focus,
  & > ${ItemTitleContent} {
    outline: none;
  }

  [data-dimension='m'] & {
    min-height: 38px;
    ${typography['Subtitle/Subtitle 3']}
  }
`;

const ItemWrapper = styled.div<{ opened?: boolean; disabled?: boolean }>`
  border-bottom: 1px solid ${(p) => p.theme.color.background.tertiary};
  color: ${(p) => p.theme.color.text.primary};
  & > ${ItemTitle} ${Chevron} {
    transform: ${(p) => (p.opened ? 'rotate(180deg)' : 'rotate(0deg)')};
  }
  ${({ disabled }) => disabled && 'pointer-events: none;'}
`;

const ItemContent = styled.div<{ contentMaxHeight: number | string }>`
  box-sizing: border-box;
  overflow-y: auto;
  max-height: ${(p) => p.contentMaxHeight};
  padding: 4px 16px 16px 16px;
  ${typography['Body/Body 1 Long']}
  [data-dimension='m'] & {
    padding: 4px 16px 10px 16px;
    ${typography['Body/Body 2 Long']}
  }
`;

const AccordionWrapper = styled.div<{ dimension?: Dimension }>`
  position: relative;
  & > ${ItemWrapper}:first-child {
    & ${ItemTitleContent} {
      padding: ${({ dimension }) => (dimension === 'l' ? '15px 16px' : '9px 16px')};
    }
    border-top: 1px solid ${({ theme }) => theme.color.background.tertiary};
  }
`;

export interface AccordionItemProps extends Omit<React.HTMLAttributes<HTMLButtonElement>, 'onClick' | 'title'> {
  /** Заголовок компонента */
  title: React.ReactNode;
  /** дефолтное (изначальное) состояние компонента (раскрыт/свернут) при неконтролируемом режиме работы */
  defaultExpanded?: boolean;
  /** состояние компонента (раскрыт/свернут) при контролируемом режиме работы */
  expanded?: boolean;
  /** Колбек на клик по компоненту */
  onClick?: (title: React.ReactNode, expanded: boolean, event: React.MouseEvent<HTMLButtonElement>) => void;
  /**  Устанавливает максимальную высоту, на которую открывается контент до появления вертикального скрола */
  contentMaxHeight?: number | string;
  /** Отключение компонента */
  disabled?: boolean;
}

export const AccordionItem: React.FC<AccordionItemProps> = ({
  children,
  title,
  id,
  defaultExpanded,
  expanded: userExpanded,
  onClick,
  contentMaxHeight = '100vh',
  disabled,
  ...props
}) => {
  const [expanded, setExpanded] = React.useState(defaultExpanded);
  const collapseOpened = userExpanded === undefined ? expanded : userExpanded;

  const itemId = React.useMemo(() => id || uid(), [id]);
  const ITEM_TITLE_ID = `accordion_title_${itemId}`;
  const ITEM_CONTENT_ID = `accordion_content_${itemId}`;

  const handleClick = React.useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      if (userExpanded === undefined) {
        onClick?.(title, !expanded, event);
        setExpanded(!expanded);
      } else {
        onClick?.(title, !userExpanded, event);
      }
    },
    [expanded, userExpanded, onClick, title],
  );
  return (
    <ItemWrapper opened={collapseOpened} data-disabled={disabled} disabled={disabled}>
      <ItemTitle
        onClick={disabled ? undefined : handleClick}
        role="button"
        type="button"
        aria-expanded={collapseOpened}
        aria-controls={ITEM_CONTENT_ID}
        id={ITEM_TITLE_ID}
        {...props}
      >
        <ItemTitleContent tabIndex={-1}>
          {title}
          <Chevron aria-hidden />
        </ItemTitleContent>
      </ItemTitle>
      <Collapse opened={collapseOpened} contentMaxHeight={contentMaxHeight}>
        <ItemContent
          contentMaxHeight={contentMaxHeight}
          role="region"
          aria-labelledby={ITEM_TITLE_ID}
          id={ITEM_CONTENT_ID}
        >
          {children}
        </ItemContent>
      </Collapse>
    </ItemWrapper>
  );
};

export interface AccordionProps extends React.HTMLAttributes<HTMLDivElement> {
  dimension?: Dimension;
}

export const Accordion: React.FC<AccordionProps> = ({ children, dimension = 'l', onKeyDown, ...props }) => {
  const accordionRef = React.useRef<HTMLDivElement | null>(null);
  const handleKeyDown = React.useCallback(
    (e) => {
      let focusedOption: any = ((accordionRef.current && accordionRef.current.ownerDocument) || document).activeElement;
      // если фокус находится на AccordionItem Title, то можно переключаться между accordion items с помощью ArrowDown, ArrowUp, Home, End
      if (focusedOption.id.indexOf('accordion_title') > -1 && focusedOption.tagName.toLowerCase() === 'button') {
        do {
          focusedOption = focusedOption?.parentNode;
        } while (focusedOption.parentNode !== accordionRef.current);

        const code = keyboardKey.getCode(e);

        if (code === keyboardKey.ArrowDown) {
          moveFocus(accordionRef.current, focusedOption, nextItem);
          e.preventDefault();
        } else if (code === keyboardKey.ArrowUp) {
          moveFocus(accordionRef.current, focusedOption, previousItem);
          e.preventDefault();
        } else if (code === keyboardKey.Home) {
          e.preventDefault();
          moveFocus(accordionRef.current, null, nextItem);
        } else if (code === keyboardKey.End) {
          e.preventDefault();
          moveFocus(accordionRef.current, null, previousItem);
        }
      }
      onKeyDown?.(e);
    },
    [accordionRef.current],
  );
  return (
    <AccordionWrapper
      ref={accordionRef}
      data-dimension={dimension}
      dimension={dimension}
      onKeyDown={handleKeyDown}
      {...props}
    >
      {children}
    </AccordionWrapper>
  );
};

Accordion.displayName = 'Accordion';
