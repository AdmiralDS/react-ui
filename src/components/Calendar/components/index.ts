export * from './Years';
export * from './Button';
export * from './Day';
export * from './DayNames';
export * from './Month';
export * from './Months';
export * from './Panel';
export * from './Week';
