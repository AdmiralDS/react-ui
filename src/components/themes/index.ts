export * from './light';
export * from './dark';
export type { Theme } from './common';
