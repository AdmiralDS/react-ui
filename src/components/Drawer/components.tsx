import type { FC, HTMLAttributes } from 'react';
import { createContext, useContext, useLayoutEffect, useRef } from 'react';
import styled from 'styled-components';

import { typography } from '#src/components/Typography';
import { debounce } from '#src/components/common/utils/debounce';
import { resizePaddings } from '../common/dom/resizePaddings';

const Title = styled.h5<{ $mobile: boolean; $displayCloseIcon: boolean }>`
  ${({ $mobile }) => ($mobile ? typography['Header/H6'] : typography['Header/H5'])};
  color: var(--admiral-color-Neutral_Neutral90, ${(p) => p.theme.color['Neutral/Neutral 90']});
  margin: 0;
  padding: ${({ $mobile, $displayCloseIcon }) => {
    if ($mobile) {
      return $displayCloseIcon ? '0 48px 8px 16px' : '0 16px 8px';
    }
    return $displayCloseIcon ? '0 56px 8px 24px' : '0 24px 8px';
  }};
`;

const Content = styled.div<{ $mobile: boolean }>`
  overflow-y: auto;
  outline: none;
  scrollbar-gutter: stable;
  padding-block: 8px;
  padding-inline: ${({ $mobile }) => `${$mobile ? 16 : 24}px`};
  height: 100%;
`;

const ButtonPanel = styled.div<{ $mobile: boolean }>`
  display: flex;
  flex-direction: ${({ $mobile }) => ($mobile ? 'column-reverse' : 'row-reverse')};
  padding: ${({ $mobile }) => ($mobile ? '16px 16px 0' : '16px 24px 0')};

  & > button {
    margin: ${({ $mobile }) => ($mobile ? '0 0 16px 0' : '0 16px 0 0')};
    ${({ $mobile }) => $mobile && 'width: 100%;'}
  }

  & > button:first-child {
    margin: 0;
  }
`;

export const DrawerContext = createContext({ mobile: false, displayCloseIcon: true } as {
  mobile: boolean;
  displayCloseIcon: boolean;
});

export const DrawerTitle: FC<HTMLAttributes<HTMLHeadingElement>> = ({ children, ...props }) => {
  const { mobile, displayCloseIcon } = useContext(DrawerContext);
  const asProp = mobile ? 'h6' : 'h5';
  return (
    <Title $mobile={mobile} $displayCloseIcon={displayCloseIcon} as={asProp} {...props}>
      {children}
    </Title>
  );
};

export const DrawerContent: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  const contentRef = useRef<HTMLDivElement | null>(null);
  const mobile = useContext(DrawerContext).mobile;

  useLayoutEffect(() => {
    const node = contentRef.current;
    if (node) {
      resizePaddings(node);
      const resizeObserver = new ResizeObserver(debounce(() => resizePaddings(node), 250));
      resizeObserver.observe(node);
      return () => {
        resizeObserver.unobserve(node);
      };
    }
  }, []);

  return (
    <Content tabIndex={-1} ref={contentRef} $mobile={mobile} {...props}>
      {children}
    </Content>
  );
};

export const DrawerButtonPanel: FC<HTMLAttributes<HTMLDivElement>> = ({ children, ...props }) => {
  const mobile = useContext(DrawerContext).mobile;
  return (
    <ButtonPanel $mobile={mobile} {...props}>
      {children}
    </ButtonPanel>
  );
};
