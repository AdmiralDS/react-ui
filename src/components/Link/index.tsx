import { TYPOGRAPHY } from '#src/components/Typography';
import * as React from 'react';
import styled, { css } from 'styled-components';

export interface LinkOwnProps {
  /** Вид ссылки */
  appearance?: AppearanceLink;
  /** Отключение ссылки */
  disabled?: boolean;
  /** Размер ссылки */
  dimension?: DimensionLink;
}

export type AppearanceLink = 'primary' | 'secondary';
export type DimensionLink = 'm' | 's';

export const styleDisabledMixin = css<{ disabled?: boolean }>`
  color: ${({ theme }) => theme.color.text.tertiary};
  & *[fill^='#'] {
    fill: ${({ theme }) => theme.color.text.tertiary};
  }
  user-select: none;
`;

export const styleColorMixin = css<{ appearance?: AppearanceLink }>`
  color: ${({ theme, appearance }) =>
    appearance === 'secondary' ? theme.color.text.primary : theme.color.basic.primary};
  & *[fill^='#'] {
    fill: ${({ theme, appearance }) =>
      appearance === 'secondary' ? theme.color.text.secondary : theme.color.basic.primary};
  }
`;

export const styleMixin = css`
  &:hover {
    color: ${({ theme }) => theme.color.basic.hover};
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.basic.hover};
    }
  }
  &:focus {
    color: ${({ theme }) => theme.color.basic.hover};
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.basic.hover};
    }
    outline-color: ${({ theme }) => theme.color.basic.hover};
  }
  &:active {
    color: ${({ theme }) => theme.color.basic.press};
    & *[fill^='#'] {
      fill: ${({ theme }) => theme.color.basic.press};
    }
  }
`;

export const styleTextMixin = css<{ dimension?: DimensionLink }>`
  font-size: ${({ dimension }) => (dimension === 'm' ? 16 : 14)}px;
  font-weight: normal;
  line-height: ${({ dimension }) => (dimension === 'm' ? 24 : 20)}px;
  font-family: ${TYPOGRAPHY.fontFamily};
`;

export const LinkCssMixin = css<LinkOwnProps>`
  cursor: ${(props) => (props.disabled ? 'default' : 'pointer')};
  display: flex;
  pointer-events: ${(props) => (props.disabled ? 'none' : 'auto')};
  align-items: center;
  width: fit-content;
  text-decoration: none;
  position: relative;
  ${styleMixin};
  ${styleColorMixin};
  ${styleTextMixin};
  ${(props) => (props.disabled ? styleDisabledMixin : '')}
`;

export const LinkComponent = styled.a<LinkOwnProps>`
  ${LinkCssMixin}
`;

export interface LinkProps extends LinkOwnProps, React.AnchorHTMLAttributes<HTMLAnchorElement> {}

export const Link = React.forwardRef<HTMLAnchorElement, LinkProps>(
  ({ disabled = false, appearance = 'primary', dimension = 'm', ...props }, ref) => {
    return (
      <LinkComponent
        dimension={dimension}
        ref={ref}
        tabIndex={disabled ? -1 : props.tabIndex}
        disabled={disabled}
        appearance={appearance}
        {...props}
      />
    );
  },
);

Link.displayName = 'Link';
