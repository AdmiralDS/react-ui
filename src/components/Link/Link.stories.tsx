import React from 'react';
import styled, { ThemeProvider } from 'styled-components';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';
import { Link } from '#src/components/Link';
import { T } from '#src/components/T';
import { ReactComponent as ArrowLeftOutline } from '@admiral-ds/icons/build/system/ArrowLeftOutline.svg';
import { ReactComponent as ArrowRightOutline } from '@admiral-ds/icons/build/system/ArrowRightOutline.svg';
import { Theme } from '#src/components/themes';

export default {
  title: 'Admiral-2.1/Link',
  decorators: [withDesign],
  component: Link,
  parameters: {
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=37%3A8489',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/EGEGZsx8WhdxpmFKu8J41G/Admiral-2.1-UI-Kit?node-id=37%3A8620',
      },
    ],
  },
  argTypes: {
    target: {
      control: {
        type: 'select',
        options: ['_parent', '_self', '_top', '_blank'],
      },
    },
    appearance: {
      control: {
        type: 'radio',
        options: ['primary', 'secondary'],
      },
    },
    dimension: {
      control: {
        type: 'radio',
        options: ['m', 's'],
      },
    },
    disabled: {
      control: { type: 'boolean' },
    },
    themeBorderKind: {
      control: {
        type: 'radio',
        options: ['Border radius 0', 'Border radius 2', 'Border radius 4', 'Border radius 8'],
      },
    },
  },
} as ComponentMeta<typeof Link>;

const Separator = styled.div`
  margin-top: 10px;
`;

const Devider = styled.div`
  width: 10px;
`;

const LinkPrimary: ComponentStory<typeof Link> = (args) => {
  function swapBorder(theme: Theme): Theme {
    theme.shape.borderRadiusKind = (args as any).themeBorderKind || theme.shape.borderRadiusKind;
    return theme;
  }

  return (
    <ThemeProvider theme={swapBorder}>
      <T font="Body/Body 1 Long" as="div">
        Dimension - M
      </T>
      <Link appearance={'primary'} href="" onClick={(e) => e.preventDefault()}>
        Link
      </Link>
      <Separator />
      <T font="Body/Body 1 Long" as="div">
        Dimension - S
      </T>
      <Link appearance={'primary'} dimension="s" href="" onClick={(e) => e.preventDefault()}>
        Link
      </Link>
    </ThemeProvider>
  );
};

const LinkSecondary: ComponentStory<typeof Link> = () => (
  <>
    <T font="Body/Body 1 Long" as="div">
      Dimension - M
    </T>
    <Link appearance={'secondary'} href="" onClick={(e) => e.preventDefault()}>
      Link
    </Link>
    <Separator />
    <T font="Body/Body 1 Long" as="div">
      Dimension - S
    </T>
    <Link appearance={'secondary'} dimension="s" href="" onClick={(e) => e.preventDefault()}>
      Link
    </Link>
  </>
);

const LinkWithIcon: ComponentStory<typeof Link> = () => {
  return (
    <>
      <T font="Body/Body 1 Long" as="div">
        Link with icon. Dimension - m
      </T>
      <Separator />
      <Link appearance={'primary'} href="" onClick={(e) => e.preventDefault()}>
        <ArrowLeftOutline width={24} />
        <Devider />
        Link
      </Link>
      <Separator />
      <T font="Body/Body 1 Long" as="div">
        Link with icon. Dimension - s
      </T>
      <Separator />
      <Link appearance={'secondary'} dimension="s" href="" onClick={(e) => e.preventDefault()}>
        Link
        <Devider />
        <ArrowRightOutline width={20} />
      </Link>
    </>
  );
};

const LinkDefault: ComponentStory<typeof Link> = ({ children, ...args }) => <Link {...args}>{children}</Link>;
export const Primary = LinkPrimary.bind({});
export const Secondary = LinkSecondary.bind({});
export const IconLink = LinkWithIcon.bind({});
export const Playground = LinkDefault.bind({});

Playground.args = {
  children: 'Link',
  href: 'http://localhost:6006/?path=/story/example-link--playground',
};
