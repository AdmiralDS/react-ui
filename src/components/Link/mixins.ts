import { css } from 'styled-components';
import type { AppearanceLink, Dimension } from '#src/components/Link/LinkComponent';
import { typography } from '#src/components/Typography';
import { mediumGroupBorderRadius } from '#src/components/themes';

export const styleDisabledMixin = css`
  color: var(--admiral-color-Neutral_Neutral30, ${(p) => p.theme.color['Neutral/Neutral 30']});
  & *[fill^='#'] {
    fill: var(--admiral-color-Neutral_Neutral30, ${(p) => p.theme.color['Neutral/Neutral 30']});
  }
  user-select: none;
`;

export const styleColorMixin = css<{ $appearance?: AppearanceLink }>`
  color: ${({ theme, $appearance }) =>
    $appearance === 'secondary'
      ? `var(--admiral-color-Neutral_Neutral90, ${theme.color['Neutral/Neutral 90']})`
      : `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`};
  & *[fill^='#'] {
    fill: ${({ theme, $appearance }) =>
      $appearance === 'secondary'
        ? `var(--admiral-color-Neutral_Neutral90, ${theme.color['Neutral/Neutral 90']})`
        : `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`};
  }
`;

export const styleMixin = css`
  border-radius: var(--admiral-border-radius-Medium, ${(p) => mediumGroupBorderRadius(p.theme.shape)});
`;

export const styleEventMixin = css`
  &:hover {
    color: var(--admiral-color-Primary_Primary70, ${(p) => p.theme.color['Primary/Primary 70']});
    & *[fill^='#'] {
      fill: var(--admiral-color-Primary_Primary70, ${(p) => p.theme.color['Primary/Primary 70']});
    }
  }
  &:focus {
    color: var(--admiral-color-Primary_Primary70, ${(p) => p.theme.color['Primary/Primary 70']});
    & *[fill^='#'] {
      fill: var(--admiral-color-Primary_Primary70, ${(p) => p.theme.color['Primary/Primary 70']});
    }
    outline-color: var(--admiral-color-Primary_Primary70, ${(p) => p.theme.color['Primary/Primary 70']});
  }
  &:active {
    color: var(--admiral-color-Primary_Primary80, ${(p) => p.theme.color['Primary/Primary 80']});
    & *[fill^='#'] {
      fill: var(--admiral-color-Primary_Primary80, ${(p) => p.theme.color['Primary/Primary 80']});
    }
  }

  &:focus-visible {
    outline-offset: 2px;
    outline: var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']}) solid 2px;
  }
`;

export const styleTextMixin = css<{ $dimension?: Dimension }>`
  ${({ $dimension }) => ($dimension === 's' ? typography['Body/Body 2 Long'] : typography['Body/Body 1 Long'])}
`;
