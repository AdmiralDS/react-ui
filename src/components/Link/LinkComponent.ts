import styled from 'styled-components';
import { AnchorHTMLAttributes } from 'react';
import { LIGHT_THEME as DEFAULT_THEME } from './../themes';
import { styleColorMixin, styleDisabledMixin, styleMixin, styleTextMixin } from './mixins';

export type AppearanceLink = 'primary' | 'secondary';
export type Dimension = 'm' | 's';

type LinkComponent = AnchorHTMLAttributes<HTMLAnchorElement> & {
  disabled: boolean;
  appearance: AppearanceLink;
  dimension: Dimension;
};

export const LinkComponent = styled.a<LinkComponent>`
  cursor: ${(props) => (props.disabled ? 'default' : 'pointer')};
  display: flex;
  pointer-events: ${(props) => (props.disabled ? 'none' : 'auto')};
  align-items: center;
  width: fit-content;
  text-decoration: none;
  position: relative;
  ${styleMixin};
  ${styleColorMixin};
  ${styleTextMixin};
  ${(props) => (props.disabled ? styleDisabledMixin : '')}
`;

LinkComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
