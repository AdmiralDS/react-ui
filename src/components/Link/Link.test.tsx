import { LIGHT_THEME } from '#src/components/themes';
import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Link } from '../Link';

const linkText = 'Link TextInput';

describe('Link', () => {
  test('should render component', () => {
    const { container } = render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Link href="#">{linkText} </Link>
      </ThemeProvider>,
    );
    expect(container).toMatchSnapshot();
  });
  test('renders Link component', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Link href="#">{linkText} </Link>
      </ThemeProvider>,
    );
    expect(screen.getByRole('link')).toHaveTextContent(linkText);
  });

  test('should focus on component when user on press key tab', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Link href="#" data-testid="test-element">
          {linkText}
        </Link>
      </ThemeProvider>,
    );
    expect(document.body).toHaveFocus();
    userEvent.tab();
    expect(screen.getByTestId('test-element')).toHaveFocus();
  });

  test('should dont focus component when user on press key tab', () => {
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Link href="#" disabled>
          {linkText}
        </Link>
      </ThemeProvider>,
    );
    expect(document.body).toHaveFocus();
    userEvent.tab();
    expect(document.body).toHaveFocus();
  });

  test('should call onClick when user clicks on component', () => {
    const handleClick = jest.fn();
    render(
      <ThemeProvider theme={LIGHT_THEME}>
        <Link href="#" onClick={handleClick}>
          {linkText}
        </Link>
      </ThemeProvider>,
    );
    fireEvent.click(screen.getByText(linkText));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });
});
