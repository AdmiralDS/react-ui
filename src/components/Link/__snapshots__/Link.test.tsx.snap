// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should render component 1`] = `
.c0 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  pointer-events: auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  -webkit-text-decoration: none;
  text-decoration: none;
  position: relative;
  color: #0D69F2;
  font-size: 16px;
  font-weight: normal;
  line-height: 24px;
  font-family: 'VTB Group UI',sans-serif;
}

.c0:hover {
  color: #0A54C2;
}

.c0:hover *[fill^='#'] {
  fill: #0A54C2;
}

.c0:focus {
  color: #0A54C2;
  outline-color: #0A54C2;
}

.c0:focus *[fill^='#'] {
  fill: #0A54C2;
}

.c0:active {
  color: #0949AA;
}

.c0:active *[fill^='#'] {
  fill: #0949AA;
}

.c0 *[fill^='#'] {
  fill: #0D69F2;
}

<div>
  <a
    class="c0"
    href="#"
  >
    Link TextInput
     
  </a>
</div>
`;
